base/shared/src/main/scala/scalaz/data/Cofree.scala

```diff
@@ -0,0 +1,22 @@
+package scalaz
+package data
+
+import scalaz.control.{Coinductive, Inf}
+
+trait CofreeModule {
+  type Cofree[F[_], A]
+
+  def runCofree[F[_], A](f: Cofree[F, A]): F[data.Cofree[F, A]]
+
+  def wrapCofree[F[_], A](a: =>A)(f: =>F[data.Cofree[F, A]]): Cofree[F, A]
+}
+
+private[data] object CofreeImpl extends CofreeModule {
+  type Cofree[F[_], A] = (Coinductive[A], Coinductive[F[data.Cofree[F, A]]])
```

Comment:
Some would say that using a type constructor (`Coinductive`) as its own type is bad practice. Don't you really want `Inf` instead of `Coinductive`?

Reply:
In talks with @edmundnoble last night we've decided to implement it it in terms of `Fix` and `EnvT` intead a la 

```scala

trait CofreeModule {
  type Cofree[F[_], A]

  def runCofree[F[_], A](f: Cofree[F, A]): (A, F[Cofree[F, A]])

  def wrapCofree[F[_], A](a: =>A)(f: =>F[Fix[EnvT[A, F, ?]]]): Cofree[F, A]
}

private[data] object CofreeImpl extends CofreeModule {
  final class EnvT[A, T[_], X](val run: (A, T[X]))

  type Cofree[F[_], A] = Fix[EnvT[A, F, ?]]

  def runCofree[F[_], A](f: Cofree[F, A]): (A, F[Cofree[F, A]]) =
    Fix.unfix[EnvT[A, F, ?]](f).run

  def wrapCofree[F[_], A](a: =>A)(f: => F[Cofree[F, A]]): Cofree[F, A] =
    Fix.fix[EnvT[A, F, ?]](new EnvT((a, f)))
}

```